patches/server/0925-Paper-Plugins.patch

```diff
@@ -0,0 +1,5591 @@
+new file mode 100644
+index 0000000000000000000000000000000000000000..6a30e697c3391b9cef814f33e5eb3f7502727df1
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/plugin/provider/service/entrypoint/storage/ordered/BootstrapProviderStorage.java
+@@ -0,0 +1,76 @@
++package io.papermc.paper.plugin.provider.service.entrypoint.storage.ordered;
++
++import io.papermc.paper.plugin.DependencyContext;
++import io.papermc.paper.plugin.DependencyContextHolder;
++import io.papermc.paper.plugin.bootstrap.PluginBootstrap;
++import io.papermc.paper.plugin.bootstrap.PluginBootstrapContext;
++import io.papermc.paper.plugin.bootstrap.PluginBootstrapContextImpl;
++import io.papermc.paper.plugin.provider.PluginProvider;
++import io.papermc.paper.plugin.provider.ProviderStatus;
++import io.papermc.paper.plugin.provider.ProviderStatusHolder;
++import io.papermc.paper.plugin.provider.service.entrypoint.storage.ordered.strategy.ModernPluginLoadingStrategy;
++import io.papermc.paper.plugin.provider.service.entrypoint.storage.ordered.strategy.PluginGraphCycleException;
++import net.minecraft.server.Main;
++
++import java.util.ArrayList;
++import java.util.List;
++import java.util.logging.Level;
++import java.util.logging.Logger;
++
++public class BootstrapProviderStorage extends OrderedProviderStorage<PluginBootstrap> {
++
++    private static final Logger LOGGER = Logger.getLogger("BootstrapProviderStorage");
++
++    public BootstrapProviderStorage() {
++        super(new ModernPluginLoadingStrategy<>(new ProviderLoader<>() {
++            @Override
++            public void applyContext(PluginProvider<PluginBootstrap> provider, DependencyContext dependencyContext) {
++                if (provider instanceof DependencyContextHolder contextHolder) {
++                    contextHolder.setContext(dependencyContext);
++                }
++            }
++
++            @Override
++            public boolean load(PluginProvider<PluginBootstrap> provider, PluginBootstrap provided) {
++                try {
++                    PluginBootstrapContext context = PluginBootstrapContextImpl.of(provider, Main.pluginDirectory);
++                    provided.boostrap(context);
++                    return true;
++                } catch (Exception e) {
++                    provider.getLogger().log(Level.SEVERE, "Failed to run bootstrapper for %s. This plugin will not be loaded.".formatted(provider.getSource()), e);
++                    if (provider instanceof ProviderStatusHolder statusHolder) {
++                        statusHolder.setStatus(ProviderStatus.ERRORED);
++                    }
++                    return false;
++                }
++            }
++        }));
++    }
++
++    @Override
++    public void processProvided(PluginBootstrap provided) {
++    }
```

Comment:
I think we'll need to store PluginBootstrap instances somewhere, (PluginBootstrapInstanceManager?) because I tried this PR out on my plugin network and there I needed to check if all PluginBootstraps which depend on one main PluginBootstraps and which were supposed to load at a server beginning have already loaded. (I needed such thing due to a complex initialization system in my plugin network) So, with such manager I'd get full class names of PluginsBootstrap instances, (if they exist in this manager, then they're loaded) then I'd get registered providers in BootstrapProviderStorage, get configuration and filtered them by the dependency, mapped to a value of a "bootstrapper" field, and checked if the first collection contains entries from the second one. But for now I can't do such check...

Reply:
Bootstraps that depend on others should correctly load before them. Bootstraps are stored in the server main "entry point" storage.

Reply:
So, after the discord discussion, what do you think, should .getPluginBootstrappers() (a method returning all loaded bootstrappers) be added somewhere is the API? Yeah, my case was too overengineered and too specific, but I think if there will be cases which also would need such method I think it should be added...